src/analysis/vector/geometry_checker/qgsvectorlayerfeaturepool.cpp

```diff
@@ -108,6 +108,9 @@ bool QgsVectorLayerFeaturePool::addFeatures( QgsFeatureList &features, QgsFeatur
 
 void QgsVectorLayerFeaturePool::updateFeature( QgsFeature &feature )
 {
+  QgsFeature origFeature;
+  getFeature( feature.id(), origFeature );
```

Comment:
I may be getting confused here, but I think there's still a risk of index corruption in this scenario:

1. Feature A is retrieved (version "1"), it is cached in the pool. Bounds are used in the spatial index.
2. More features are fetched, cache fills up, feature A is evicted from cache
3. Some external caller changes the geometry on feature A in the vector layer or data provider, now there's version 2 of feature A.
4. A call to updateFeature for feature A is made. getFeature won't find it in the pool's cache (it was evicted), so it's retrieved from the layer. We get feature A version 2.
5. Version 2 is passed to refreshCache, and the index is corrupted because we try to remove different bounds (version 2, whereas the cache is using v1)

I don't see any alternative but keeping a local map/hash of the feature bounds which were used in the index.

Reply:
You're right, I did not expect that the spatial index could evict a feature... It is an index, not strictly a cache, isn't it?

Reply:
It won't get evicted from the spatial index, but will from mFeatureCache.

Reply:
Of course 🙈   obviously.

Okay, what about "synchronizing" the cache and the index?

In fact, in the end of step 4. inside the index, we have both feature A version 2 and feature A version 1.

We could fix this: when getFeature doesn't find a cached feature, we can proactively delete any pre-existing corresponding feature in the index (see new commit).

If the feature was not present in the index, it does nearly nothing, otherwise it removes the possibly out-of-date feature from the index, to re-add the possibly same or up-to-date feature.